### PR TITLE
SONARAZDO-418 Ensure systematic .NET team dogfooding before Marketpla…

### PR DESCRIPTION
**Pull request by lucas-paulger-sonarsource**
…ce releases

Please review our [contribution guidelines](https://github.com/SonarSource/sonar-scanner-azdo/blob/master/docs/contributing.md).

And please ensure your pull request adheres to the following guidelines: 

- [ ] Please explain your motives to contribute this change: what problem you are trying to fix, what improvement you are trying to make
- [ ] Use the following formatting style: [SonarSource/sonar-developer-toolset](https://github.com/SonarSource/sonar-developer-toolset#code-style)
- [ ] Provide a unit test for any code you changed
- [ ] If there is a [JIRA](https://jira.sonarsource.com/browse/SONARAZDO) ticket available, please make your commits and pull request start with the ticket ID (SONARAZDO-XXXX)

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -16,6 +16,8 @@ variables:
     value: true
   - name: NPM_CONFIG_registry
     value: $(ARTIFACTORY_URL)/api/npm/npm
+  - name: isMain
+    value: $[eq(variables['Build.SourceBranch'], 'refs/heads/feature/lp/create-dogfooding-extension')] # replace with refs/heads/master
```
Comment (lucas-paulger-sonarsource) [inline azure-pipelines.yml:20]: after validating this check works, I will change to `refs/head/master`

File: azure-pipelines.yml [hunk 2]
```diff
@@ -185,3 +189,45 @@ stages:
             inputs:
               command: "test"
               projects: "its/runner/IntegrationTests.csproj"
+
+  - stage: publish_dogfood
+    displayName: "Publish dogfood extension"
+    condition: and(succeeded(), eq(variables['isMain'], true))
```
Comment (7PH) [inline azure-pipelines.yml:195]: NP / disclaimer: not tested

```suggestion
    condition: and(succeeded(), variables['isMain'])
```

File: azure-pipelines.yml [hunk 3]
```diff
@@ -185,3 +189,45 @@ stages:
...
+    jobs:
+      - job: "publish_dogfood"
+        variables:
+          publisher: $(SonarSourceITPublisher)
+          scExtensionVersion: $[stageDependencies.build.npminstall.outputs['versions.SC_VERSION']]
+          deps: $[convertToJson(dependencies)] # to debug if dependencies are correct
```
Comment (7PH) [inline azure-pipelines.yml:201]: I don't understand where this comes from?

File: azure-pipelines.yml [hunk 4]
```diff
@@ -185,3 +189,45 @@ stages:
...
               projects: "its/runner/IntegrationTests.csproj"
+
+  - stage: publish_dogfood
+    displayName: "Publish dogfood extension"
+    condition: and(succeeded(), eq(variables['isMain'], true))
+    jobs:
+      - job: "publish_dogfood"
+        variables:
+          publisher: $(SonarSourceITPublisher)
+          scExtensionVersion: $[stageDependencies.build.npminstall.outputs['versions.SC_VERSION']]
```
Comment (7PH) [inline azure-pipelines.yml:200]: I think it fails because this variable comes from another stage, wdyt?

File: gulpfile.js [hunk 5]
```diff
@@ -497,7 +497,7 @@ gulp.task("ci:azure:get-extensions-version", (done) => {
     console.log(`Fetched ext version ${version} for ${extension}`);
     console.log(`##vso[task.setvariable variable=${extensionPrefix}_EXT_NAME]${extensionManifest}`);
     console.log(
-      `##vso[task.setvariable variable=${extensionPrefix}_VERSION;isOutput=true]${version}`,
+      `##vso[task.setvariable variable=${extensionPrefix}_VERSION;isOutput=true;isreadonly=true]${version}`,
```
Comment (lucas-paulger-sonarsource) [inline gulpfile.js:500]: I was thinking it was getting overridden between stages, but as I found out, only the last stage is passed to the `stageDependencies`. I can omit this but in theory it is read only 🤷🏻 
https://learn.microsoft.com/en-us/azure/devops/pipelines/process/set-variables-scripts?view=azure-devops&tabs=bash#set-variable-properties

File: azure-pipelines.yml [hunk 6]
```diff
@@ -185,3 +187,40 @@ stages:
...
+              connectTo: "VsTeam"
+              connectedServiceName: "AzDO Extension - Marketplace - Dogfood"
+              fileType: "vsix"
+              vsixFile: "$(System.ArtifactsDirectory)/extensions/sonar-scanner-azdo-$(scExtensionVersion)-sonarcloud.vsix"
+              extensionName: "[Dogfood] SonarCloud"
+              extensionId: "sonarcloud-dogfood"
+              updateTasksVersion: false
+              updateTasksId: false
+              extensionVisibility: "private"
+              shareWith: "sonarsource"
```
[the remark this answers is not in the document]
Comment (7PH) [reply azure-pipelines.yml:232]: I agree, like it inline now TBH

**Review by 7PH — APPROVED** [review]
We don't have the permission to install the extension on the `SonarSource` organization, so that'll be another thing to check

LGTM ✅
image

File: azure-pipelines.yml (hunk 2 above)
Comment (lucas-paulger-sonarsource) [reply azure-pipelines.yml:198]: I'll keep it as is in that it clearly states what we expect (boolean true)
examples shown here
https://learn.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#stage-to-stage-dependencies

File: azure-pipelines.yml (hunk 1 above)
Comment (lucas-paulger-sonarsource) [reply azure-pipelines.yml]: addressed and works

**Pull request merged**